csrc/scheduler/registry_utils.cpp

```diff
@@ -572,59 +572,96 @@ bool SchedulerTopologyChecker::hasNonNormalizePostReductionBCast(
 
-        auto forward_pairwise_root_map = PairwiseRootDomainMap(
-            forward_running_producer, forward_running_consumer);
-        auto forward_p2c_root_map =
-            forward_pairwise_root_map.mapProducerToConsumer();
-
-        // These are the ids we will have to resolve. As we resolve them we'll
-        // remove them from this vector. If this vector ends up empty, then
-        // we've resolved everything we need to. This is a pair so as we
-        // traverse we can map the id through the traversal. The first entry
-        // in the pair will be the original id so we can reset it if it's not
-        // resolved before the next traversal. The second ID will be
-        // propagated as we map the IDs through the backward traversal.
-        std::vector<std::pair<IterDomain*, IterDomain*>> ids_to_resolve;
-
-        // Check if any TensorViews have a resolved broadcast
-        for (auto entry : forward_p2c_root_map) {
-          auto p_id = entry.first;
-          auto c_id = entry.second;
-          if (p_id->isBroadcast() && !c_id->isBroadcast()) {
-            ids_to_resolve.emplace_back(c_id, c_id);
+          // Check if any TensorViews have a resolved broadcast
+          for (auto entry : forward_p2c_root_map) {
+            auto p_id = entry.first;
+            auto c_id = entry.second;
+            if (p_id->isBroadcast() && !c_id->isBroadcast()) {
+              ids_to_resolve.emplace_back(c_id, c_id);
+            }
+          }
+          if (ids_to_resolve.empty()) {
+            continue;
           }
         }
 
-        if (ids_to_resolve.empty()) {
-          continue;
+        // At this point, we have some ids to resovle.
+        // Check if we can move forward further along the use chain.
+        // This helps gathering more usage info to resolve this id to avoid
+        // unnecessary segmentation when the reduction input iteration domain is
+        // used after this tensor. See PostReductionBroadcastCheck test and
+        // issue-2146. We can move forward when:
+        // (1) The usage chain is not empty.
+        // (2) The consumer is not a broadcastOp, as broadcastOp introduces a
+        //     new broadcast axis and require a new check, see
+        //     FusionReshapePersistentShmoo.
+        // (3) At least one id is mapped, otherwise can't do backward traversal.
+        if (!forward_tv_dep_chain.empty()) {
+          auto tmp_producer = forward_running_consumer;
+          auto tmp_consumer = forward_tv_dep_chain.front();
+          if (!tmp_consumer->definition()->isA<BroadcastOp>()) {
+            auto forward_pairwise_root_map =
+                PairwiseRootDomainMap(tmp_producer, tmp_consumer);
+            auto forward_p2c_root_map =
+                forward_pairwise_root_map.mapProducerToConsumer();
+            bool at_leat_one_id_mapped = false;
+            for (size_t entry_i = ids_to_resolve.size(); entry_i > 0;
```

Comment:
Is there any reason traversing from the back?

Reply:
The only reason I can think of is we are removing element from the vector, so the safest way is to vist backward. I added a case with multiple bcast ids and did some experiment with forward iterating, it also works fine as long as the loop condition is correctly set, e.g. use 
```
size_t num = ids_to_resolve.size();
for (size_t entry_i = 0; entry_i < num;entry_i++)
```
instead of 
`for (size_t entry_i = 0; entry_i < ids_to_resolve.size();entry_i++)`

Reply:
Where is the removal?

Reply:
[Here](https://github.com/NVIDIA/Fuser/blob/ad7ea4466612cca4b2e8bf9b27e6efeb2bdaa3b1/csrc/scheduler/registry_utils.cpp#L693):
```
                    // If mapped, and producer is a producer of a reduction,
                    // we can resolve this id
                    ids_to_resolve.erase(
                        ids_to_resolve.begin() + (int64_t)entry_i - 1);
```